sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala

```diff
@@ -0,0 +1,356 @@
+    case _ => Nil
+  }
+
+  /**
+   * Adds the expression to this data structure recursively. Stops if a matching expression
+   * is found. That is, if `expr` has already been added, its children are not added.
+   */
+  def addExprTree(
+      expr: Expression,
+      map: mutable.HashMap[GpuExpressionEquals, GpuExpressionStats] = equivalenceMap): Unit = {
+    val skip = expr.isInstanceOf[LeafExpression] ||
+      expr.isInstanceOf[GpuLeafExpression] ||
+      expr.isInstanceOf[GpuUnevaluable] ||
+      (expr.isInstanceOf[GpuExpression] && expr.asInstanceOf[GpuExpression].hasSideEffects) ||
+      // `LambdaVariable` is usually used as a loop variable, which can't be evaluated ahead of the
+      // loop. So we can't evaluate sub-expressions containing `LambdaVariable` at the beginning.
+      expr.find(_.isInstanceOf[LambdaVariable]).isDefined ||
+      // `PlanExpression` wraps query plan. To compare query plans of `PlanExpression` on executor,
+      // can cause error like NPE.
+      (expr.find(_.isInstanceOf[PlanExpression[_]]).isDefined && TaskContext.get != null)
+
+    if (!skip && !addExprToMap(expr, map)) {
+      childrenToRecurse(expr).foreach(addExprTree(_, map))
+      commonChildrenToRecurse(expr).filter(_.nonEmpty).foreach(addCommonExprs(_, map))
+    }
+  }
+
+  /**
+   * Returns the state of the given expression in the `equivalenceMap`. Returns None if there is no
+   * equivalent expressions.
+   */
+  def getExprState(e: Expression): Option[GpuExpressionStats] = {
+    equivalenceMap.get(GpuExpressionEquals(e))
+  }
+
+  // Exposed for testing.
+  private[sql] def getAllExprStates(count: Int = 0): Seq[GpuExpressionStats] = {
+    equivalenceMap.values.filter(_.useCount > count).toSeq.sortBy(_.height)
+  }
+
+  /**
+   * Returns a sequence of expressions that more than one equivalent expressions.
+   */
+  def getCommonSubexpressions: Seq[Expression] = {
+    getAllExprStates(1).map(_.expr)
+  }
+
+  /**
+   * Returns the state of the data structure as a string. If `all` is false, skips sets of
+   * equivalent expressions with cardinality 1.
+   */
+  def debugString(all: Boolean = false): String = {
+    val sb = new java.lang.StringBuilder()
+    sb.append("GPU Equivalent expressions:\n")
+    equivalenceMap.values.filter(stats => all || stats.useCount > 1).foreach { stats =>
+      sb.append("  ").append(s"${stats.expr}: useCount = ${stats.useCount}").append('\n')
+    }
+    sb.toString()
+  }
+}
+
+object GpuEquivalentExpressions extends Logging {
+  def logExprTree(expr : Expression): Unit = {
+    logWarning("nodeName: " + expr.nodeName + "\n" + expr.treeString)
+  }
+
+  def logExprs(title: String, exprs: Seq[Expression]): Unit = {
+    if (!exprs.isEmpty) {
+      logWarning("Start: " + title)
+      exprs.foreach(logExprTree)
+      logWarning("  End: " + title)
+    }
+  }
+
+  def logExprTiers(title: String, tiers: Seq[Seq[Expression]]): Unit = {
```

Comment:
Are these needed still?

Reply:
I am still using them for debugging, but I will remove before merging.

Reply:
Removed.